### ISSUE TITLE
Include an "Editor" role in the standard install profile

### ISSUE THREAD

**Issue by sutibun** [backdrop/backdrop-issues#1285]
Looking at the default roles, you'd think there are no content creators in the Bd world.

![roles](https://cloud.githubusercontent.com/assets/14132746/10214868/c66f13ae-67e1-11e5-96c5-5fcb6d6b7673.png)

Some roles to possibly add. Not sure what is the best name
1. contributor, writer, author?
   - give this role the ability to create and edit their own content.
   - No admin stuff.
   - Can unpublish their own content but not delete.
2. editor
   - edit any content regardless of authorship.
   - has the power to unpublish and delete any content.
   - power to disable/freeze any contributor account (role 1 above) from submitting new content. Basically, if contributor is being bad or incompetent, the content will remain but since account closed, they won't be able to work on the site anymore.
3. guest poster
   - can only create unpublished content

Dunno about the last 2 roles but throwing them out there

---
~PR by @stpaultim: https://github.com/backdrop/backdrop/pull/2830~
Updated PR by @quicksketch: https://github.com/backdrop/backdrop/pull/2837

**Comment by docwilmot** [comment]
I think this is a great idea:

> Joomla has three user groups:
Authors can submit articles.
Editors can submit new articles and edit existing articles.
Publishers can submit new articles, edit existing articles, and publish articles.

Perhaps similar?

**Comment by sutibun** [comment]
I looked a lil' bit into their [docs](https://docs.joomla.org/J2.5:How_permissions_work_in_Joomla!_2.5) as I didn't understand what "submit" meant

> Author 
> - Registered Group privileges 
> - Create new articles but cannot publish articles 
> - Edit articles they own 
> - View special content 
>
> Editor 
> - Author Group privileges 
> - Can edit all articles, even those that are not published
>
> Publisher 
> - Editor Group privileges 
> - Can publish articles

It looks similar to what I wrote except they don't give the editor the right to publish a piece, just edit. They leave that to the "Publisher". Somehow that sounds wrong.

**I'd still merge editor and publisher into one but I am not 100% confident.** Maybe I am missing something? I'd like to hear what others think or what has been their experience.

**Comment by klonos** [comment]
What I find myself doing in 99% of the websites I build with user/groups is to...

1. leave myself as "admin" (unless the firm has an in-house system administrator, in which case I add them as one or simply hand them the credentials for it).
2. leave the "anonymous" role as is (but always thought that this should be renamed to "site visitors" or plain "visitors" - "anonymous" seems too technical for the end users).
3. leave the "authenticated" role also as is (but wished to be able to rename it to "registered users" or just "registered" because it also makes more sense).
4. add a "site owner" role. This role has some specific super-privileges over the simple site users (whatever the site owner has requested to be able to do like edit all nodes, add/remove users and assign/unassign roles), but still much more limited than the privileges the "admin" role has.
5. add a "contributor" or "author" role. This one usually does what the equivalent Wordpress group does.

...usually Editor/Publisher additional privileges are permitted to the "site owner". So far there was no request to delegate these to a separate role. I guess site owners of small site still like to have control over things and fair enough.

**Comment by klonos** [comment]
...there are cases where the site owner is actually computer/web literate and there is no need for a "*site owner*" role or they even specifically request that the admin role is assigned to them  once the project is finished. These are rare cases though.

**Comment by sutibun** [comment]
## Renaming roles

> leave the "anonymous" role as is (but always thought that this should be renamed to "site visitor" or plain "visitor" or even better "guest" - "anonymous" seems too technical for the end users ...not to mention that there is the risk to be confused/misinterpreted for other things).

To me, "Site visitor" & "visitor" are a lil' too broad and doesn't quite seem to define the role.

Maybe I'm too biased here but I've always liked "Anonymous" and I understood it immediately when I first came to Drupal. Maybe we can rename to "Anonymous visitors" so it's less ambiguous?

I like "guest" too. It's broad but still indicates the type of person... someone who is passing by but doesn't belong (not registered)

> leave the "authenticated" role also as is (but wished to be able to rename it to "registered users" or just "registered", because it also makes more sense).

Perfect! "Registered users" definitely makes way more sense. *I* should've thought of that!

I always had to pause and think what "authenticated" meant. Not to mention, *that* is way too technical.

## Contributor + Editor/Publisher role

> add a "contributor" or "author" role. This one usually does what the equivalent Wordpress group does
>
> ...usually Editor/Publisher additional privileges are permitted to the "site owner". So far there I had no request to delegate these to a separate role. I guess site owners of small site still like to have control over things (fair enough).

Interesting that there is no need for an editor role. Would you then suggest no "Editor" role as a default? (e.g. just Anonymous + Authenticated + Contributor/Author + Admin?)

**I can imagine the need for an "editor" role but still not sure about a publisher.** The "publisher" role seems iffy to me... something a large organization would need.

The not-too-fiction scenario I am thinking is a small non-profit wanting to publish site content. They have volunteers but, obviously, to ensure the site doesn't get defaced, they put someone in charge just to make sure the content passes mustard.

What about you @docwilmot? How are most sites you set up like? What roles do you usually need?

**Comment by klonos** [comment]
> Would you then suggest no "Editor" and no "Publisher" role as a Bd default? (e.g. just Anonymous + Authenticated + Contributor/Author + Admin?)

I would say that for small and medium-sized businesses, the following roles would suffice:

- guest
- registered user
- editor/contributor
- site owner
- site admin

For some websites, the site owner doesn't want to have to do anything with the site administration tasks. They usually are happy to be given a "*editor*"/"*contributor*" role. It's just that usually, along the way, they keep asking to be given "special" privileges and they simply do not understand that the way the underlying system is set, if they ("editor" role) are given certain privileges, so are the rest of the users that belong to the "editor" group. That's when (though not specifically asked to) as a workaround, I create the "site owner" role.

The reason for separate "site owner" and "site admin" groups is simply because you just cannot (and perhaps *should* not ever) give admin/super-user permissions to people that could accidentally "break" the site. Think of the site admin role as reserved for the system admin or the web designer - similar case as the linux root if you like.

One thing that has always bothered me though, is the fact that new privileges (say some added by contrib modules or after a core upgrade) are not always auto-assigned to the site admin. I think that we should special-case the site admin role and have it have all privileges (existing ones + future additions). I simply cannot imagine use cases where a site admin would not want to have full privileges (could be wrong though) - it is a given I believe and we should treat it as a "sensible default". That would also remove the respective column from the "Permissions" page (less clutter - better UX if you ask me). Separate issue perhaps?

Of course, one could think that perhaps the "*guest*" user is not needed either (as in not being an actual role that is). We could simply regard a user as guest if they have no other specific role. Not sure if that is guesswork and whether it imposes any security risks, but although it could simplify things, on the other hand there are cases where site guest visitors are given some "small" privileges. If that user role does not already exist out of the box, we are in essence asking site developers to create it. So removing this role would defeat the purpose of improving UX.

**Comment by jenlampton** [comment]
I don't know about this.  Most of my sites have two roles. Developer (me) and editor/admin (client).  Sometimes the editor/admin uses "Authenticated".

I don't know that we can make any assumptions about which (or even how many) roles are in use without real data.  Let's add this to the update tracker and postpone for a while :) https://github.com/backdrop-ops/backdropcms.org/issues/99

**Comment by klonos** [comment]
FTR, the GovCMS distro has a "Content author" role, a "Content approver" (same permissions as the authors + moderation), a "Site builder" and a "Site editor". The "administrator" (user 1) is reserved and locked down.

In GovCMS8 the "Site builder" and "Site editor" roles have been merged into a single "Site editor" role.

**Comment by herbdool** [comment]
The article talked about *editors* not *authors*. Which do you mean @klonos?

**Comment by klonos** [comment]
"Content editors" and "Content authors" are the same roles to me @herbdool. GovCMS has "Content author" (often referred to as "Content editor"), and "Content approver". The difference between them is that "Content approvers" can also publish content, which is otherwise created as draft by default.

**Comment by pixelite** [comment]
I think the labels "content editor", "author", and "content manager" tend to change from site to site (which is the reason that we don't already have out-of-the-box roles). To me, a content editor has more permissions than an author, usually also permission to edit other people's content and to publish things, and perhaps permission to manage things like menus and taxonomy. That seems to reflect the tasks that self-ascribed content editors do regularly in the survey we conducted. I like the idea of providing a role like this, but it's certainly something that a site builder would need to adjust for each site.

**Comment by klonos** [comment]
I agree @pixelite ...depending on each site and the size of the team(s) managing a website, chances are that the exact permissions each of these "personas" will have varies. It think that we should try to be slightly "restrictive", while at the same time providing enough to have a useful set of tasks shown in the admin bar for such users. Then it will certainly be up to the site admins to make this role (or roles if we decide to add more than one) more permissive.

The risk of adding such roles is very minimal, as in very small sites chances are that these roles will be left unassigned 😉

**Comment by klonos** [comment]
I have come across some related issues on d.org:

[Add a content editor role to Drupal's standard profile](https://www.drupal.org/project/ideas/issues/2703367)

> In user story phrasing, "A module developer can assign permissions to a content editor role."
>
>Drupal is a content management system. Having a content editor role is not a ridiculous stretch. With this role and the configuration it allows, out-of-the-box experience for people using Drupal core and any modules that choose to use progressive enhancement if this particular role is present will be much improved.

[Rename 'Authenticated user' to 'logged-in user' and 'anonymous user' to 'visitor'](https://www.drupal.org/project/drupal/issues/479708)

> **Problem/Motivation**
> Partly inspired by http://www.d7ux.org/roles-the-admin-header but I've been thinking about it for a while.
> 
> 1. Authenticated user and anonymous user are somewhat drupalisms - or at least I've never seen them used elsewhere
> 2. It's been shown in UB user testing and anecdotally that lots of people don't understand giving a permission to auth users gives it to every logged in user. See [#78941: Usability: Auto-check permissions if "authenticated" has them](https://www.drupal.org/project/drupal/issues/78941).
> 3. Anonymous users is terrible - these are visitors to the site, authenticated users can be 'anonymous' too.
> 
> So I suggest just 'logged in users' and 'visitors'.
> 
> **Proposed resolution**
> "Anonymous user" role label changes to "Visitor" (singular)
> "Authenticated user" role label changes to "Logged-in user" (singular)

[Remove hardcoded anonymous and authenticated user roles](https://www.drupal.org/project/drupal/issues/468768)

> For a long time I've been wondering if this was possible, and eventually I decided to try writing a patch for it :)
> 
> The attached patch removes Drupal's hardcoded anonymous and authenticated user roles and replaces them with a simple system where any role on a site can be auto-assigned to users (a) before they log in, and (b) right after they have registered for an account.
> 
> The patch is basically functional and has the following advantages:
> 
> 1. The "anonymous user" and "authenticated user" roles that Drupal currently has are weird and confusing. Why? Because "anonymous" and "authenticated" represent states of interaction with a computer, but most other roles represent categories of people (think "Editor", "Administrator", etc). Things would be much clearer if the roles that users are automatically assigned work the same way as other roles on the site. For example, in this initial patch the default install profile starts off with anonymous users being assigned to a "Guest" role and newly-registered users being granted a "Member" role, but the expert profile doesn't start with any roles at all, and sites are free to use whatever roles are meaningful to them.
> 2. The patch makes it possible to auto-assign the same role for both (a) and (b) above, if you want unauthenticated and authenticated users on your site to share the same base set of permissions. (No more clicking two checkboxes in a row on the permissions page!)
> 3. Some sites have no need for "authenticated users" at all (for example, single user blogs). Currently, Drupal forces that role on them, but this patch means it won't anymore.
> 4. If your site registration policies get more strict (for example, if you previously allowed open registration but then later switched to requiring email verification), this patch allows you to easily separate your users into "before" and "after" by switching the role that is auto-assigned to newly registered users. For example, you could create a new role called "Verified Member" and auto-assign that; you can then safely grant this new role more permissions knowing that anyone who gets it must have gone through more verification steps during registration than users who got the older "Member" role.
> 5. Any patch which involves deleting as much ugly, special-case code as this one does can't be a bad thing :)

**Comment by klonos** [comment]
It seems that they are going with "Content Editor" in D8: https://www.drupal.org/project/ideas/issues/2703367 ...and they are also considering a "Content Manager" role too (proposal ATM): https://docs.google.com/document/d/1q2XO-T2rfK6lxQNzHWxb7noCYw8athAleOfnR3CWOt0

**Comment by jenlampton** [comment]
We should keep in mind that the target audience for Drupal 8 is vastly different from that of Backdrop. Many Backdrop sites won't need roles for both editors and managers in addition to administrators. Adding the "needs usage metrics" tag so we can tell what the 80% will need.

**Comment by klonos** [comment]
The preliminary results from the survey are out (166 submissions at the time of writing): https://docs.google.com/document/d/1bN6EBefKqezuZ-oGUHt4cPxBk7sidaqNGWKINX1WZ_0

> More than 84% approve of including a content editor role in Drupal core. 10.8% were neutral on the proposal, and only about 4.2% actually disagreed.

...

>The respondents were more ambivalent about adding a content manager role as a default. Those who agreed/strongly agreed totaled 44%. The neutral votes were quite high at 25.3%. Those who disagreed/strongly disagreed came in at 30.7% combined.

...

>In a couple of cases, respondents wrote in that they did not agree with adding a default role because they view its usage as too specific and not relevant to most websites. 
>
>One suggestion that seemed interesting to us to consider in the future is to ask a question during the base Drupal install on whether to add the content editor role or not rather than installing it automatically.

**Comment by klonos** [comment]
Perhaps we should consider introducing an enabled/disabled status for roles, and have any additional roles (content editor/manager) be disabled out of the box. All this in the spirit of people finding it easier to enable and then tweak existing things (rather than create them from scratch).

**Comment by jenlampton** [comment]
It's not hard to create a role. All you have to do is type in the name.

What's the point of building all this infrastructure if new riles would be
disabled out of the box? I don't see any value to adding disabled roles,
but a ton of work.

Let's also remember that Drupal's audience is not Backdrop's audience. I'm
not sure there's a majority use-case for additional roles in Backdrop.

Adding complexity to the permissions page is the last thing we want to do.
Adding **unnecessary** roles will he a major UX regression.

**Comment by laryn** [comment]
I can see value in having a role like  "Content Editor" set up and configured out of the box. For one, it may prevent people from granting the "Administrator" role more loosely than they should simply because they want to allow another account to add/edit content. (I think it's possible they won't think through what other permissions might be on by default with that role).

Sure, you could type in the name to create the role, but then you also have to go through all the permissions and figure out which ones you need/select appropriately, and you have to know to do that in the first place...

**Comment by findlabnet** [comment]
"Authenticated" versus "Anonymous" already cover 99% of typical small sites as far as I can see.

**Comment by klonos** [comment]
> What's the point of building all this infrastructure if new riles would be
disabled out of the box?

This suggestion was so to limit the number of columns shown in permission tables OOTB. We can add this, or not.

> It's not hard to create a role. All you have to do is type in the name.

For the d.org proposal, this was not simply about shipping new roles OOTB (which is indeed easy); it was also about shipping these roles with sensible permissions. That is where the value for end users comes in.

We have been talking about adopting features from the WordPress word too, because they seem to be a very popular CMS with less technical audience. Well, WP ships with 5 user roles OOTB, and "Author" and "Editor" are included:

<img width="763" alt="Screen Shot 2019-04-03 at 6 30 41 am" src="https://user-images.githubusercontent.com/2423362/55430693-7ecb6080-55da-11e9-8b99-9bcdf6f9310a.png">

> Default WordPress user roles are designed to have capabilities that fits the requirement of most websites. For example if you run a magazine site, then you can assign Editor user role to your senior staff and author user role to the junior staff. You can assign contributor user role to your guest authors and subscriber user role for your site visitors.

**Comment by jlfranklin** [comment]
> It's not hard to create a role. All you have to do is type in the name. 

I concur that additional roles should not be there out of the box by default.  Or, more accurately, in the default profile. 

> We have been talking about adopting features from the WordPress world [...] WP ships with 5 user role OOTB. 

I don’t think that we need that in the default profile, but we could create a couple targeted profiles that preconfigure a site for specific purposes, including a general blogging site and a newsletter site. Both of those would have done predefined roles demonstrating how to set up Backdrop for content creation and publishing. 

Extra profiles will add some release time work to make sure that they still install correctly, a burden we’d have to accept.

**Comment by klonos** [comment]
> Extra profiles will add some release time work to make sure that they still install correctly, a burden we’d have to accept.

Or... add the additional roles (even if "disabled"), so that we don't have the burden 😅 

All joking aside, in my head, this is equivalent to implementing #3530 / #3531 / #3648. Creating a layout is easy, but specifying the correct context is not. Having layouts where node, user, and term contexts are already set, is a UX+. Many sites may never use them, that's why the thought was to have these disabled (layouts already support this).

**Comment by herbdool** [comment]
I find the comparison to WordPress compelling. I'm leaning towards content editor role.

I wonder if there was a similar discussion in WP.

**Comment by jenlampton** [comment]
When it comes to user roles, there's not only the UX nightmare that is the permissions matrix, but also the security risk that comes along with that. Layouts don't present the same kind of risk, and offer more of a reward.

I also find it interesting that WP comes with so many roles, considering most WP sites have only a single user. But I also wonder what kind of risk is involved with adding more WP roles.

This is the difference between roles and layouts...

> (layouts already support this)

For Backdrop we'd need to build in disabled role support (and all the security risk that comes with it) for new functionality that would not benefit the majority. We'd be adding risk to 100% of sites, but adding reward, to maybe 10%? 

The telemetry we collect on role names will be interesting.

**Comment by olafgrabienski** [comment]
>> We have been talking about adopting features from the WordPress word too, because they seem to be a very popular CMS with less technical audience. Well, WP ships with 5 user roles OOTB, and "Author" and "Editor" are included

> I find the comparison to WordPress compelling. I'm leaning towards content editor role.
> I wonder if there was a similar discussion in WP. 

I don't know of the discussion in WordPress, but an important difference is that WP (without extra plugins) doesn't allow to manage any permissions of the roles. That might be the reason for having so many roles out of the box.

Apart from the comparison to WordPress, I'm also leaning towards an editor role with sensible defaults for Backdrop. To discuss and test the defaults, might be a lot of work, though.

**Comment by stpaultim** [comment]
I hear the concerns of @jenlampton, but I also lean towards having a site editor role with sensible defaults. I don't know if it's 80% of the time, but quite often the person who builds or configures the site is not the person that works with it day to day. Too often site editors get left with the admin role and access to tons of configuration options that they will never need. 

I believe that many Drupal site owners or editors get frustrated dealing with the full admin menu because no one ever takes the time to set up sensible permissions for someone with basic site editing duties. I try to do this for all my clients, but frequently inherit sites where is has not been done. 

When I ask clients if they prefer a simpler admin menu without so many things that they will never use, they ALWAYS say 'yes."

**Comment by Graham-72** [comment]
I agree. If I am providing a site that someone else will edit I always provide a site editor role with limited permissions compared with admin.

**Comment by jenlampton** [comment]
I could be convinced to add one more role. Editor is the best candidate.

**Comment by BWPanda** [comment]
I also vote for a new 'Editor' role, enabled OOTB, that allows those users to manage all the content side of things (add/edit/delete/etc.). On developing a new site, the developer can then add/remove permissions as necessary from that role before handing it over.

Re. naming, I really like single-word role names.

**Comment by klonos** [comment]
> Re. naming, I really like single-word role names.

I think that plain "Editor" has been discussed in Drupal, but was rejected in favor of the more specific "Content editor". The reason IIRC was possible disambiguation in use cases like online magazines where "Editor" means different a thing.

Personally, so long as we provide some "lesser" admin role, with basic content editing permissions, I don't have a specific strong preference. I'm mostly interested in the OOTB UX, and feature parity with Wordpress/D8.

**Comment by klonos** [comment]
...FTR, recent activity in https://www.drupal.org/project/ideas/issues/2703367

> This was discussed during the UX meeting on 2019-04-16 (https://www.youtube.com/watch?v=uyilEZAZGNI), and @webchick was there. Her feedback was:
> - We already have enough research here to support the need of this new editor role, so we can move forward.
> - Adding this new role won't affect current installations, so it's safe on terms of backwards compatibility.
> 
> She also suggested to create a meta issue to discuss what else we could do with the editor role in core. So @antonellasev has created this 2 issues:
> 
> - [#3059984](https://www.drupal.org/project/drupal/issues/3059984): Add new “Content Editor” role to Standard Profile
> - [#3059986](https://www.drupal.org/project/drupal/issues/3059986): [META] New features for the new “Content Editor” role in Standard Profile

**Comment by BWPanda** [comment]
> The intention IIRC was disambiguation in use cases like online magazines where "Editor" has a different meaning.

My personal, unsupported opinion is that 'Editor' would suit 80% of Backdrop installs just fine, and the odd online magazine that needs something different can simply rename it :slightly_smiling_face:

**Comment by klonos** [comment]
> My personal, unsupported opinion is that 'Editor' would suit 80% of Backdrop installs just fine, and the odd online magazine that needs something different can simply rename it 🙂

Totally fair 😄

**Comment by stpaultim** [comment]
I'm not sure if we have sufficient consensus on this, but I went ahead and created a PR to give folks a change to weight in. I have the sense that most of the comments support adding at least (or no more than) one new role. I called the new role "Editor" and added the permissions listed below. 

### Unauthenticated User

- access comments
- access content
- use text format filtered_html

### Authenticated User

- post comments
- skip comment approval
- edit own comments
- upload editor images

### Editor

- access dashboard
- access content overview
- view own unpublished content
- view any unpublished content
- view revisions
- revert revisions
- delete revisions
- create page content
- edit own page content
- edit any page content
- delete own page content
- delete any page content
- create post content
- edit own post content
- edit any post content
- delete own post content
- delete any post content
- access administration bar
- view the administration theme

### Administrator

- administer blocks
- administer comments
- synchronize configuration
- administer dashboard
- administer fields
- administer view modes
- access file overview
- manage files
- delete files
- administer filters
- upload editor files
- use text format full_html
- administer image styles
- use installer browser
- administer menu
- bypass node access
- administer content types
- administer nodes
- view hidden paths
- administer url aliases
- create url aliases
- administer path patterns
- notify of path changes
- administer redirects
- administer search
- search content
- use advanced search
- administer modules
- administer site configuration
- administer themes
- administer software updates
- access administration pages
- access site in maintenance mode
- access site reports
- block IP addresses
- flush caches
- administer taxonomy
- edit terms in tags
- delete terms in tags
- administer permissions
- administer users
- assign roles
- administer account settings
- access user profiles
- change own username
- cancel account
- select account cancellation method,
- administer views
- access all views
- access contextual links
- administer layouts

https://github.com/backdrop/backdrop/pull/2830

**Comment by klonos** [comment]
Thanks @stpaultim 👍 ...I have left some feedback in the PR. You used the same `$admin_role` variable for the new editor role, and as a result, the role has not been created in the PR sandbox. Once you've fixed things, you may need to refresh the sandbox to test if the role actually gets created upon installation (by closing the PR, waiting for a couple fo mins for the sandbox to be removed, then reopening the PR).

**Comment by klonos** [comment]
...I have no feedback at the moment re the set of permissions assigned to the new role, but they seem fine. Perhaps check what our Drupal brethren are doing in https://www.drupal.org/project/ideas/issues/2703367 😉

**Comment by klonos** [comment]
Interesting! ...the author of https://www.drupal.org/project/tasty_backend has chimed in on that d.org issue (I was not aware pf that project till now):

> Tasty Backend provides a ready made administration experience for day-to-day content administrators.
> 
> Most admin users won't, and shouldn't, have full access to the Drupal administration backend. They need to concentrate on tasks like adding and managing content, taxonomies, menus, and possibly other users. They do not need access to configure views, install/configure modules, or any other full administration privileges.

...the same person mentions:

> The most important thing I've probably noticed in creating TB, is where core falls short in permissions. This is where TB relies on contrib. For content, TB requires the following modules:
> 
> - [Menu Admin Per Menu](https://www.drupal.org/project/menu_admin_per_menu)
> - [Override Node Options](https://www.drupal.org/project/override_node_options)
> - [View Unpublished](https://www.drupal.org/project/view_unpublished)

I think that we are in a slightly better position in Backdrop core, since we've implemented #897 (view_unpublished).

We already have an issue about implementing most of the functionality of the override_node_options contrib module in Backdrop core (#815), and one for implementing some/most of menu_admin_per_menu (#384). These tickets are old, but there have been multiple comments that suggest that there's consensus re the usefulness of their features OOTB in Backdrop core.

**Comment by klonos** [comment]
@stpaultim here's the d.org issue about the actual implementation, and the most recent comment re which permissions should be assigned to that role: https://www.drupal.org/project/drupal/issues/3059984#comment-13175084

**Comment by stpaultim** [comment]
Thanks. Good stuff. I went ahead and added a few more permissions to this role, like clearing the cache and managing taxonomies. Curious to see what other folks think.

**Comment by BWPanda** [comment]
My 2c:

- Wonder if Editor should be able to create custom blocks...?
- Add comment admin
- Add edit own comments
- Add use contextual links
- Add the 3 File permissions (or at least the first one)
- Add upload files through editor
- Add full HTML
- Add administer content
- Add view hidden paths
- Add create and edit URL aliases
- Add notify of URL alias changes
- Remove use site in maintenance mode

These are obviously my recommendations regarding default permissions for the Editor role. And I find it easier to view the permissions page, so if the sandbox has been changed at all, some of my suggestions might not be accurate...

**Comment by BWPanda** [comment]
Also, if it helps you can copy what I do here https://github.com/PackWeb/pw_profile/blob/master/install/pw_profile.users.inc re. using `array_merge` so you don't have to duplicate permissions inntge PR code. Each role just lists the permissions they have different from the previous one.

**Comment by olafgrabienski** [comment]
Of @BWPanda's list, I'd add at least "Use contextual links" and "Create and edit URL aliases". Personally, I'm quite cautious with most of the other new suggestions but they make perfect sense if the default character of the editor role should be rather generous.

**Comment by klonos** [comment]
> ...at least "Use contextual links" and "Create and edit URL aliases"

I second that 👍

**Comment by olafgrabienski** [comment]
> I was looking for a quick win on a visible feature that might get into 14.0.0 and this issue (...) seemed like it might be pretty easy.

I'm all for promoting this feature, and it might seem easy to have a first result for the next minor release. I'd suggest however to discuss first if the new editor role should be rather generous or rather restrictive. If the latter, we need probably more time to fix issues with permissions which are not differentiated enough, e.g. Menus, Node options, Taxonomy.

**Comment by klonos** [comment]
Yeah, it is surely worth mentioning during the next dev meeting.

**Comment by stpaultim** [comment]
The consensus at the DEV meeting today seemed to be to lean towards a more restrictive set of permissions initially. We can change permissions at any time in the future, since this only effects new sites and will never effect an existing site. 

I asked about the need for tests. @quicksketch said that we could probably get away with only a test that the role has been successfully implemented, since existing tests for permissions should probably cover most everything else in the PR. 

I will try to review the suggested permissions by tomorrow and then we can bikeshed the specifics later and/or debate adding them in a future release. If anyone is able to help with the tests, great. Otherwise, I'll try to copy some existing tests and reuse them. I've never written tests before (other than behat tests).

**Comment by quicksketch** [comment]
@stpaultim I made a new PR that includes a basic test for the new user role. Much to my surprise, Backdrop didn't actually run tests provided by install profiles (or themes). I fixed that as well. I left all permissions the same as you had previously.

PR: https://github.com/backdrop/backdrop/pull/2837

**Comment by klonos** [comment]
Thanks @quicksketch 👍 ...I have left some code suggestions in the PR (typos and coding standards mostly). I think that this looks good. We can bikeshed re which permissions over the next couple of weeks before the final release 😉

**Comment by quicksketch** [comment]
> Backdrop didn't actually run tests provided by install profiles (or themes). I fixed that as well.

Well that was a bad idea. Other tests broke when I did that. So for now I've moved the testing of the Standard Profile into SimpleTest module, where we have a bunch of other system-wide tests.

**Comment by klonos** [comment]
> Well that was a bad idea. ...for now I've moved the testing of the Standard Profile into SimpleTest module ...

Follow-up here: #4008

**Comment by stpaultim** [comment]
Sounds like this is moving along fine without me, which is good, because I won't really be available until Tuesday after feature freeze. Agreed that we still have time to bikeshed specific permissions and I think we've decided that they can be changed/updated at anytime since they only effect new installs.

**Comment by quicksketch** [comment]
Okay now that we've got tests 100% passing again, I've merged https://github.com/backdrop/backdrop/pull/2837, which still maintains the exact set of permissions @stpaultim put in his PR.

Please create follow-ups for any changes in permissions. Thanks to the *many* people who weighed in on this!
